### DISCUSSION TITLE
Send interval

### DISCUSSION THREAD

**Discussion by chrissli** [JoelBender/bacpypes#514] [Q&A] [answered]
Hey there!
I've been trying to work with bacpypes to send some sensordata via bacnet.
I tested it with Yabe and I can see my sensors and values and I can subscribe to them, but they refresh only every 60 seconds.
Is there an easy way to change the  refresh rate e.g. via the BACpypes.ini or some other way?

Thanks in advance,
Chris.

**Comment by JoelBender** [comment]
What services are you using to send the data?  Are you "pushing" it out via Write-Property to another device, or is the other device relying on COV notifications via a subscription?

If you are using something like [this](https://github.com/JoelBender/BACpypes3/blob/main/samples/cov-server.py) then the notification rate is going to be a combination of the ramp rate, sleep times, and the COV_Increment.

**Comment by chrissli** [comment]
Sorry for my late reply, christmas hit hard this year. Might be a bit late, but also Happy New Year!

To be honest, it's a while ago that I started this and it worked just fine, since I never had the need to update more often.
As it seems my implementation is based on an older example.
So here is what I did:
```
def main():
	global tsensorList
	
	args = ConfigArgumentParser(description=__doc__).parse_args()

	this_device = LocalDeviceObject(
		objectName=devicename,
		objectIdentifier=int(args.ini.objectidentifier),
		maxApduLengthAccepted=int(args.ini.maxapdulengthaccepted),
		segmentationSupported=args.ini.segmentationsupported,
		vendorIdentifier=int(args.ini.vendoridentifier),
		vendorName="MyTemperature",
	)


	# this is only a wrapper
	# class MyApplication(BIPSimpleApplication, ChangeOfValueServices):
	this_application = MyApplication(this_device, args.ini.address)
		
	# adding objects
	
	moduleCounter = 1	
	tsensorList = {}
	threadList = {}


	# this is actually in a loop based on a list of temperatureSensors, therefor the "identifier"-variable
	tsensorList[identifier] = AnalogValueObject(
		objectIdentifier = ('analogValue',moduleCounter),
		objectName = identifier,
		presentValue = '',
		statusFlags = [0, 0, 0, 0],
		covIncrement = 1.0,
		outOfService = False,
		eventState = 'normal',
		units = 'noUnits',
		
		# comment out to show all
		propertyList = ArrayOfPropertyIdentifier(['statusFlags','units','presentValue','outOfService','eventState'])
	)

	moduleCounter += 1
	this_application.add_object(properties[ moduleIdentifier ])	
	threadList[identifier] = AnalogueValueThread(5, identifier)
	deferred( threadList[identifier].start )
	

	this_covclient = SubscribeCOVRequest()	
	run()
```

I hope I did not forget anything in this summary, the complete code contains too much unrelated stuff.

So what I'm doing is more comparable to the "custom-server.py" sample, I guess. 
I'm keeping the AnalogValueObject in a global list and start an AnalogueValueThread 
( which is just a Thread that keeps the AnalogValueObject.presentValue updated with the readings from my temperature-sensor[s] every 5 seconds )

I guess I have to adjust something in the wrapper for BIPSimpleApplication?
If needed I can redo this with asyncio instead of threads, but I assume that's not the problem, right?

Sorry to bother you

**Comment by JoelBender** [comment, marked as answer]
Happy New Year! :partying_face: 

There are problems with using threads, the "ownership" of an object belongs to the thread that created it, so when you create your AnalogValueObject in the main thread and update in some other one wires can get crossed and it's hard to track down.

Rather than fixing this with the "legacy" BACpypes this is _much_ simpler in BACpypes3.  Start with the [cov-server.py](https://github.com/JoelBender/BACpypes3/blob/main/samples/cov-server.py) sample application and replace the `ramp()` function with one that reads the temperature sensor and updates the value.

**Reply by JoelBender** [reply]
LOL - Apparently I mentioned this earlier in this discussion.  At least I'm consistent! :zany_face:

**Comment by chrissli** [comment]
Yeah, I thought so.

I tried the cov-server.py and saw the covIncrement is directly connected to the update rate whereas, even if I strip down my programm to the essentials ( without threads ) the update rate won't change with the covIncrement.

I'm just gonna recreate it based on the cov-server.py


Anyway, thank you for your kind effort and the great work you did there!

**Reply by chrissli** [reply]
Hey there,

sorry for bothering you again.
I have a little follow-up question about this, maybe you can lead me in the the right direction.
Based on the cov-server.py I tried to add more than one AnalogValueObject to the app.
Now if I loop through my "sources" and add an object each to the app and call the ramp-function on that object,
the app will block on the first object, since the ramp-function is "await"ed, but is an infinite loop.

So what I did was this:

```
if data:
	counter = 1	
	loop = asyncio.get_running_loop()
	for source in data:
		# create a custom object
		try:
			print(source)
			analog_value_object = AnalogValueObject(
				objectIdentifier=("analog-value", counter),
				objectName=source,
				presentValue=data[source],
				covIncrement=1.0,
			)
			counter += 1
	
			app.add_object(analog_value_object)
		except:
			print('Ignored: ',sys.exc_info())

		# ramp up and down
		loop.ensure_future(ramp(analog_value_object))
```

Which leads me to that error: AttributeError: 'IPv4DatagramServer' object has no attribute 'local_transport'

Is there one of your samples dealing with adding multiple objects?
Or am I just stupid here and the preferred way is to  call the ramp function only once and let it deal with all existing objects instead of calling one ramp per object?

Sorry for asking stupid questions, but I never worked with asyncio before

**Reply by chrissli** [reply]
So, I came to the solution to store my analog_object_values in a list and give that as a parameter to the ramp-function, so that one function loops over all objects and refreshes their values.
Works great.

I think the local transport error was a mistake by me
